File: 0279-perfect-squares/0279-perfect-squares.cpp
class Solution {
public:
    long long sol(vector<int>&v,int i,int n,vector<vector<int>>&dp){
       
        if(n==0)return 0;
        if(i<0)return INT_MAX-2;
        if(dp[i][n]!=-1)return dp[i][n];
        long long t=INT_MAX,n_t=INT_MAX;
        if(v[i]<=n)t=sol(v,i,n%v[i],dp)+n/v[i];
        n_t=sol(v,i-1,n,dp);
        return min(t,n_t);
        
    }
    int numSquares(int n) {
        int s=sqrt(n);
       
        vector<int>v(s);
        vector<vector<int>>dp(s,vector<int>(n+1,-1));
        for(int i=0;i<s;i++){v[i]=pow((i+1),2);}
        return sol(v,s-1,n,dp);
    }
};

File: 0279-perfect-squares/0279-perfect-squares.java
import java.util.Arrays;

class Solution {
    public long sol(int[] v, int i, int n, int[][] dp) {
        if (n == 0) return 0;
        if (i < 0) return Integer.MAX_VALUE - 2;
        if (dp[i][n] != -1) return dp[i][n];
        
        long t = Integer.MAX_VALUE;
        long n_t = Integer.MAX_VALUE;
        
        if (v[i] <= n) t = sol(v, i, n % v[i], dp) + n / v[i];
        n_t = sol(v, i - 1, n, dp);
        
        return Math.min(t, n_t);
    }

    public int numSquares(int n) {
        int s = (int) Math.sqrt(n);
        
        int[] v = new int[s];
        int[][] dp = new int[s][n + 1];
        
        for (int[] row : dp) Arrays.fill(row, -1);
        
        for (int i = 0; i < s; i++) {
            v[i] = (int) Math.pow(i + 1, 2);
        }
        
        return (int) sol(v, s - 1, n, dp);
    }
}
